### ISSUE TITLE
Add capability to process DNADerivedData to `occurrences` module

### ISSUE THREAD

**Issue by ayushanand18** [iobis/pyobis#98]
## Overview
The `pyobis` package in its current state does not support fetching `DNADerivedData` using the `occurrences` module. 

I believe there can be two ways to implement it:
+ add it as a parameter in the `search` function like `MeasurementOrFacts`
+ create a new function for it say `dna( self, **args, **kwargs )` and enable necessary processing and utility methods since `search` function is already heavy with so many features.

**Comment by 7yl4r** [comment]
[DNADerivedData is a Darwin Core extension](https://rs.gbif.org/extensions.html#http://rs.gbif.org/terms/1.0/DNADerivedData) in the same way MeasurementOrFacts is, so I think they should be treated the same. I agree that occurrences.search is already too heavy. 

----------------------

Let's come at the problem from the other direction and think about the ideal user interface without technical considerations. This is what felt intuitive to me:

```python
occurrences = pyobis.search.occurrences(
    scientificname='Mola mola',
    required_extensions=["MoF","DNA"]
)
```

However, we have already decided to separate the query-building from the query-executing, and I think that was good choice. Maybe we could aim something like this instead:

```python
occurrences = pyobis.OccQuery(
    scientificname='Mola mola',
    required_extensions=["MoF","DNA"]
).execute()
```

This is nice because it positions the code to potentially generalize to handle other DwC providers as well:

```python
occurrences = pyobis.OccQuery(
    scientificname='Mola mola',
    required_extensions=["MoF","DNA"]
).execute(
    data_providers=["OBIS", "GBIF"]
)
```

-----------------------------

So, with this in mind, I am thinking we begin adding new features to the `OccQuery` class rather than into the `search` function. Assuming we implement DNA in this way and save the other changes for later this will look like:

```python
occurrences = pyobis.OccQuery(
    required_extensions=["DNA"]
).search(
    scientificname='Mola mola',
    MeasurementOrFact=True
)
```

Thoughts?

**Comment by ayushanand18** [comment]
I agree that the ways you listed are much intuitive and user-friendly than the current project. 

But I had two questions to ask:
+ If we start adding features to the `OccQuery` class, then there are some methods (almost all except `search` only) that don't need `extensions` as parameter.
+ We could not pass parameters directly to `OccQuery class` since all methods had different parameters. For example, while `search` and `getpoints`, etc need `scientificname`, `taxonid`, etc. but `get` function requires `occurrence UUID` and nothing else.   I couldn't find an efficient way to solve this.

**Comment by 7yl4r** [comment]
I think the object-oriented response to this is that we have multiple subclasses of `OccQuery`:
* `OccGetQuery`
* `OccSearchQuery`

Both have shared attributes and methods under the parent class `OccQuery`
* query execution returns Occurrences
* shared basic API parameters like baseURL

---------------

The object-oriented approach can be a bit pedantic, but it would work here. However, our focus should remain on keeping the intuitiveness of the user-interface first.. Is the object-oriented approach starting to over-complicate the UI? 

Maybe a good compromise is to use OO programming under the hood and provide convenience methods to wrap that complexity for casual users. By this I mean little functions like this example:

```python
# this function available as pyobis.occurrences.search()
define search(scientific_name):
    return pyobis.OccSearchQuery(
        scientificname=scientific_name
    ).execute()
```

**Comment by 7yl4r** [comment]
I think this looks good, but I have a tendency to overengineer things. @ocefpaf (and|or) @MathewBiddle : what do you think? The object-oriented programming approach is familiar to me, and wrapping things with convenience methods for casual users seems reasonable to me too. But maybe there is a more elegant approach?
